compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -0,0 +1,485 @@
+  std::string str() const;
+
+  // Contains the shape of the layout
+  DimArray shape;
+  // Contains the order of layout dims for each of the tensor dims
+  OrderArray order;
+  // Shape of the tensor when used in mma
+  std::array<int, maxTensorDims> canonicalShape;
+  int rank;
+};
+
+// MMA Layout Utilities
+enum class MMAType {
+  M16N8K16,
+  NONE,
+};
+
+enum class MMAMatrixType { AMatrix, BMatrix, CMatrix };
+
+static std::array<Dimension, 3> getMMADimensions(MMAType mmaType,
+                                                 MMAMatrixType matrixType,
+                                                 int dim) {
+  switch (mmaType) {
+    case MMAType::M16N8K16:
+      switch (matrixType) {
+        case MMAMatrixType::AMatrix:
+          if (dim == 0)
+            return {{{DimType::LaneIdY, 8},
+                     {DimType::VecIdZ, 2},
+                     {DimType::LaneIdZ, 1}}};
+          return {{{DimType::VecIdX, 2},
+                   {DimType::LaneIdX, 4},
+                   {DimType::VecIdY, 2}}};
+        case MMAMatrixType::BMatrix:
+          if (dim == 0)
+            return {{{DimType::VecIdX, 2},
+                     {DimType::LaneIdX, 4},
+                     {DimType::VecIdY, 2}}};
+          return {{{DimType::LaneIdY, 8},
+                   {DimType::LaneIdZ, 1},
+                   {DimType::VecIdZ, 1}}};
+        case MMAMatrixType::CMatrix:
+          if (dim == 0)
+            return {{{DimType::LaneIdY, 8},
+                     {DimType::VecIdY, 2},
+                     {DimType::LaneIdZ, 1}}};
+          return {{{DimType::VecIdX, 2},
+                   {DimType::LaneIdX, 4},
+                   {DimType::VecIdZ, 1}}};
+      }
+      break;
+    default:
+      return {};
+  }
+}
+
+static std::array<int, 2> getMMACanonicalShape(MMAType mmaType,
+                                               MMAMatrixType matrixType) {
+  switch (mmaType) {
+    case MMAType::M16N8K16:
+      switch (matrixType) {
+        case MMAMatrixType::AMatrix:
+          return {16, 16};
+        case MMAMatrixType::BMatrix:
+        case MMAMatrixType::CMatrix:
+          return {16, 8};
+      }
+      break;
+    default:
+      return {};
+  }
+}
+
+Layout::Layout(const DimOrderArray &dimOrder,
+               const std::array<int, maxTensorDims> &canonicalShape) {
+  assert((dimOrder.size() > 0) && (dimOrder.size() <= maxTensorDims));
+  for (int i = 0; i < maxTensorDims; i++) {
+    int j;
+    for (j = 0; j < dimOrder[i].size(); j++) {
+      Dimension dim = dimOrder[i][j];
+      order[i][j] = dim.type;
+      shape[dim.type] = dim.value;
+    }
+    // Add batch dimension to the end
+    if (i == 0) {
+      order[i][j] = DimType::Batch0;
+      shape[DimType::Batch0] = 1;
+    } else {
+      order[i][j] = DimType::Batch1;
+      shape[DimType::Batch1] = 1;
+    }
+    this->canonicalShape[i] = canonicalShape[i];
+  }
+  rank = dimOrder.size();
+}
+
+void Layout::reduceDim(int i) {
```

Comment:
this seems unused.

Reply:
yes its for the multi-dim reductions but will remove for now.